Implement audience selection panel

nfmohit opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #8157 

## Relevant technical choices

This PR implements the audience selection panel.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

nfmohit commented:
## Note to Code Reviewer

It seems the E2E test (WordPress nightly) failure isn't specific to this PR and is failing for all. I'll raise this internally too if it hasn't been done already.

hussain-t reviewed (changes requested):
Excellent work, @nfmohit. The changes look good overall. However, please consider the following points:

1. The dashboard crashes if the `getConfiguredAudiences` selector's value is `null`. This is because the default value set to the `savedItemSlugs` prop will be overridden if `null` is provided. To fix this, we can conditionally check the length of the `savedItemSlugs` array. See this line for reference:
   [SelectionPanelItems.js#L54](https://github.com/google/site-kit-wp/blob/b62d31f5455326107bd13c1084dcd5ef7a642df4/assets/js/components/SelectionPanel/SelectionPanelItems.js#L54). We can also provide a fallback empty array for the `savedItemSlugs` prop in `AudienceItems`.

![Screenshot 2024-05-30 at 9 05 26 PM](https://github.com/google/site-kit-wp/assets/24408261/1ca43c23-0630-4f68-8d47-9045509977d7)

2. Please review my comments.

3. Update the branch to the latest version and resolve any conflicts.

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/Header.js:
```diff
@@ -0,0 +1,91 @@
+	const settingsURL = useSelect( ( select ) =>
+		select( CORE_SITE ).getAdminURL( 'googlesitekit-settings' )
+	);
+	const isSavingSettings = useSelect( ( select ) =>
+		select( CORE_USER ).isSavingKeyMetricsSettings()
+	);
```
Oops, it looks like I mistakenly missed removing this after I copied over the component from Key Metrics. I've removed it, but I also added a `TODO` comment to disable the link when the audiences are being saved. This can be done as part of #8158 (which I have picked up to execute) as that's where the saving functionality will be introduced. Let me know if you think otherwise, thanks!

techanvil replied:
Thanks @hussain-t and @nfmohit. Implementing this in #8158 sounds sensible. As part of that issue we can determine whether we need an `isSavingAudienceSettings()` selector, or simply use local state if we only need to reference the saving state in one place.

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/Footer.js:
```diff
@@ -0,0 +1,103 @@
+
+	if ( selectedItemsCount < MIN_SELECTED_AUDIENCES_COUNT ) {
+		metricsLimitError = sprintf(
+			/* translators: 1: Minimum number of metrics that can be selected. 2: Number of selected metrics. */
```
Copy paste error, updated, thanks!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/Footer.js:
```diff
@@ -0,0 +1,103 @@
+		metricsLimitError = sprintf(
+			/* translators: 1: Minimum number of metrics that can be selected. 2: Number of selected metrics. */
+			__(
+				'Select at least %1$d metrics (%2$d selected)',
```
Copy-paste error, updated, thanks!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/Footer.js:
```diff
@@ -0,0 +1,103 @@
+		);
+	} else if ( selectedItemsCount > MAX_SELECTED_AUDIENCES_COUNT ) {
+		metricsLimitError = sprintf(
+			/* translators: 1: Maximum number of metrics that can be selected. 2: Number of selected metrics. */
```
Copy-paste error, updated, thanks!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/AudienceItems.js:
```diff
@@ -0,0 +1,150 @@
+		select( MODULES_ANALYTICS_4 ).getConfiguredAudiences()
+	);
+	const availableAudiences = useSelect( ( select ) => {
+		const { getResourceDataAvailabilityDate, getReport } =
+			select( MODULES_ANALYTICS_4 );
+
+		const audiences = select( MODULES_ANALYTICS_4 )
+			.getAvailableAudiences()
```
Updated, thanks!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/AudienceItems.js:
```diff
@@ -0,0 +1,150 @@
+	const savedItemSlugs = useSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).getConfiguredAudiences()
+	);
+	const availableAudiences = useSelect( ( select ) => {
+		const { getResourceDataAvailabilityDate, getReport } =
+			select( MODULES_ANALYTICS_4 );
+
+		const audiences = select( MODULES_ANALYTICS_4 )
+			.getAvailableAudiences()
+			// Filter out "Purchasers" audience if it has no data.
+			?.filter( ( { audienceSlug, name } ) => {
+				if ( 'purchasers' !== audienceSlug ) {
+					return true;
+				}
+
+				return !! getResourceDataAvailabilityDate( name, 'audience' );
+			} );
+
+		if ( undefined === audiences ) {
+			return undefined;
+		}
+
+		if ( ! audiences.length ) {
+			return [];
+		}
```
Good call! I've added a new `getConfigurableAudiences` selector so that this can be reused. Please let me know if this looks good now, thanks!

hussain-t replied:
SGTM 👍

nfmohit commented:
> Excellent work, @nfmohit. The changes look good overall. However, please consider the following points:
> 
> 1. The dashboard crashes if the `getConfiguredAudiences` selector's value is `null`. This is because the default value set to the `savedItemSlugs` prop will be overridden if `null` is provided. To fix this, we can conditionally check the length of the `savedItemSlugs` array. See this line for reference:
>    [SelectionPanelItems.js#L54](https://github.com/google/site-kit-wp/blob/b62d31f5455326107bd13c1084dcd5ef7a642df4/assets/js/components/SelectionPanel/SelectionPanelItems.js#L54). We can also provide a fallback empty array for the `savedItemSlugs` prop in `AudienceItems`.
> 2. Please review my comments.
> 3. Update the branch to the latest version and resolve any conflicts.

Thank you for the kind review, @hussain-t ! 

1. Exellent catch there. I've applied a fix for it.
2. I've addressed all your comments.
3. I've updated the branch.

Please let me know if this looks good now, thank you!

hussain-t reviewed (changes requested):
Great work. Thanks for addressing the changes, @nfmohit.

I have spotted an issue with the selected group count. As you can see in the first screencast, it takes account of the filtered-out "Purchasers" audience.


https://github.com/google/site-kit-wp/assets/24408261/708931ad-bc10-45fe-ac98-c16b053a0597


Instead, should we set the selected group using the `getConfigurableAudiences` as seen in the screencast below?

```js
const savedItemSlugs = useSelect( ( select ) =>
	select( MODULES_ANALYTICS_4 )
		.getConfigurableAudiences()
		?.map( ( { name } ) => name )
);

const onSideSheetOpen = useCallback( () => {
	setValues( AUDIENCE_SELECTION_FORM, {
		[ AUDIENCE_SELECTED ]: savedItemSlugs,
	} );
}, [ savedItemSlugs, setValues ] );
```

https://github.com/google/site-kit-wp/assets/24408261/38ad849e-b795-404f-9886-c9504e3bccc0

Or perhaps we should map the `getConfiguredAudiences` with the `getConfigurableAudiences[i].name`?

nfmohit commented:
Excellent catch, thank you Hussain! This is back with you for another look.

hussain-t reviewed (changes requested):
Thanks for the great work, @nfmohit. Overall, it functions well. However, I have a couple of suggestions for improvement:

1. The optional chaining can be safely removed from the conditions. Since the `configuredAudiences` and `configurableAudiences` arrays are already gated by the fallback empty array, the optional chaining isn't necessary.

2. In the return statement, filtering the array first before mapping would enhance efficiency. This approach avoids unnecessary iterations over the array and ensures that the mapping operation is only performed on the subset of elements that will be used.

To avoid another iteration of review, I'll go ahead and apply the changes myself.

hussain-t reviewed (approved):
Excellent work, @nfmohit. LGTM 🎖️ 

Note: I've verified that the E2E failures are unrelated to this PR.

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/AudienceItems.js:
```diff
@@ -0,0 +1,142 @@
+	};
+
+	const availableSavedItems = availableAudiences
+		?.filter( ( { name } ) => savedItemSlugs?.includes( name ) )
+		.reduce( audiencesListReducer, {} );
+
+	const availableUnsavedItems = availableAudiences
+		?.filter( ( { name } ) => ! savedItemSlugs?.includes( name ) )
```
Again, there are a couple of unneeded ops:

```suggestion
		?.filter( ( { name } ) => savedItemSlugs.includes( name ) )
		.reduce( audiencesListReducer, {} );

	const availableUnsavedItems = availableAudiences
		?.filter( ( { name } ) => ! savedItemSlugs.includes( name ) )
```

nfmohit replied:
Updated, thank you!

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/Header.js:
```diff
@@ -0,0 +1,88 @@
+	const { navigateTo } = useDispatch( CORE_LOCATION );
+
+	const onSettingsClick = useCallback(
+		() => navigateTo( `${ settingsURL }#/admin-settings` ),
```
It's worth mentioning that the requirement to scroll to the Audience Segmentation section on the settings page isn't covered here.

I do realise it could be a bit fiddly though particularly considering the Selection Panel isn't integrated yet. So, I will create a new issue to cover this point.

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/index.stories.js:
```diff
@@ -0,0 +1,121 @@
+				metrics: [ { name: 'totalUsers' } ],
+			};
+
+			const setupRegistry = ( registry ) => {
```
Please can you update the story setup to avoid this console error?

![image](https://github.com/google/site-kit-wp/assets/18395600/acce3d62-b54b-41e1-bdb0-0c64b3ba5fa3)

nfmohit replied:
Fixed, thank you!

techanvil reviewed (changes requested):
Hey @nfmohit, great work here. I have left a handful of comments, most very trivial ones which I would have fixed myself, but there a couple of slightly more consequential ones which tipped the balance to sending it back your way. Nothing major, though!

hussain-t commented on assets/js/components/SelectionPanel/SelectionPanelItems.js:
```diff
@@ -51,7 +51,7 @@ export default function SelectionPanelItems( {
 				// Split list into two sections with sub-headings for current selection and
 				// additional items if there are already saved items.
-				savedItemSlugs.length !== 0 && (
+				savedItemSlugs?.length !== 0 && (
```
Hi @techanvil, please have a look into the first observation of this [comment](https://github.com/google/site-kit-wp/pull/8751#pullrequestreview-2088773072).

techanvil replied:
Hey @hussain-t, thanks for the context here.

Doesn't this line 42 prevent `savedItemSlugs` from being `null` though?

https://github.com/google/site-kit-wp/blob/4438c8ee60b244a0d3788f3268c74ad650b655d7/assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/AudienceItems.js#L39-L42

Anyhow, if we do need to keep the operator for some reason, I would suggest that we modify it as follows to avoid a potential flicker:

```suggestion
				savedItemSlugs?.length > 0 && (
```

nfmohit commented on assets/js/components/SelectionPanel/SelectionPanelItem.js:
```diff
@@ -46,9 +48,19 @@ export default function SelectionPanelItem( {
 				value={ slug }
 			>
+				{ subtitle && (
+					<span className="googlesitekit-selection-panel-item-subtitle">
```
Updated, thank you!

nfmohit commented on assets/js/components/SelectionPanel/SelectionPanelItem.js:
```diff
@@ -46,9 +48,19 @@ export default function SelectionPanelItem( {
 				{ children }
 			</SelectionBox>
+			{ suffix && (
+				<span className="googlesitekit-selection-panel-item-suffix">
```
Updated, thank you!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/AudienceItem.js:
```diff
@@ -0,0 +1,92 @@
+			description={ description }
+			isItemSelected={ isItemSelected }
+			onCheckboxChange={ onCheckboxChange }
+			suffix={ !! userCount && numFmt( userCount ) }
```
Ah, I assumed we wouldn't want to show a zero value for some reason. Updated now, thank you!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/AudienceItems.js:
```diff
@@ -0,0 +1,142 @@
+		const { rows = [] } = report || {};
+
+		return audiences?.map( ( audience ) => {
+			const rowIndex = rows?.findIndex(
```
Updated, thank you!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/AudienceItems.js:
```diff
@@ -0,0 +1,142 @@
+
+			return {
+				...audience,
+				userCount: rows?.[ rowIndex ]?.metricValues?.[ 0 ]?.value || 0,
```
Updated, thank you!

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/AudienceItems.js:
```diff
@@ -0,0 +1,142 @@
+				description = '';
+				break;
+			case 'SITE_KIT_AUDIENCE':
+				citation = __( 'created by Site Kit', 'google-site-kit' );
```
Updated, thank you!

nfmohit commented:
Thank you for the kind review, Tom! This is now back with you for another look.

Note: I've included [this fix](https://github.com/google/site-kit-wp/pull/8751/commits/5be1183dcf5bf5736f8f3ba980c56d2a13ff1009) which makes the inner components source `savedItemSlugs` from the parent.

techanvil reviewed (approved):
LGTM, nice work @nfmohit!
